pallets/ocex/src/lib.rs

```diff
@@ -901,98 +928,108 @@ pub mod pallet {
+			let mut processed_withdrawals = vec![];
+			let mut failed_withdrawals = vec![];
 			ensure!(
 				<Withdrawals<T>>::contains_key(snapshot_id),
 				Error::<T>::InvalidWithdrawalIndex
 			);
 			// This entire block of code is put inside ensure as some of the nested functions will
 			// return Err
-			ensure!(
-				<Withdrawals<T>>::mutate(snapshot_id, |btree_map| {
-					// Get mutable reference to the withdrawals vector
-					if let Some(withdrawal_vector) = btree_map.get_mut(&account) {
-						while withdrawal_vector.len() > 0 {
-							// Perform pop operation to ensure we do not leave any withdrawal left
-							// for a double spend
-							if let Some(withdrawal) = withdrawal_vector.pop() {
-								if let Some(converted_withdrawal) = withdrawal
-									.amount
-									.saturating_mul(Decimal::from(UNIT_BALANCE))
-									.to_u128()
+			<Withdrawals<T>>::mutate(snapshot_id, |btree_map| {
+				// Get mutable reference to the withdrawals vector
+				if let Some(withdrawal_vector) = btree_map.get_mut(&account) {
+					while withdrawal_vector.len() > 0 {
+						// Perform pop operation to ensure we do not leave any withdrawal left
+						// for a double spend
+						if let Some(withdrawal) = withdrawal_vector.pop() {
+							if let Some(converted_withdrawal) = withdrawal
+								.amount
+								.saturating_mul(Decimal::from(UNIT_BALANCE))
+								.to_u128()
+							{
+								if Self::transfer_asset(
+									&Self::get_pallet_account(),
+									&withdrawal.main_account,
+									converted_withdrawal.saturated_into(),
+									withdrawal.asset,
+								)
+								.is_ok()
 								{
-									if Self::transfer_asset(
-										&Self::get_pallet_account(),
-										&withdrawal.main_account,
-										converted_withdrawal.saturated_into(),
-										withdrawal.asset,
-									)
-									.is_ok()
-									{
-										processed_withdrawals
-											.try_push(withdrawal)
-											.unwrap_or_default();
-									} else {
-										// Storing the failed withdrawals back into the storage item
-										withdrawal_vector
-											.try_push(withdrawal.clone())
-											.unwrap_or_default();
-										Self::deposit_event(Event::WithdrawalFailed(withdrawal));
-									}
+									processed_withdrawals.push(withdrawal.to_owned());
+								} else {
+									// Storing the failed withdrawals back into the storage item
+									failed_withdrawals.push(withdrawal.to_owned());
+									Self::deposit_event(Event::WithdrawalFailed(
+										withdrawal.to_owned(),
+									));
 								}
+							} else {
+								return Err(Error::<T>::InvalidWithdrawalAmount)
 							}
 						}
-						// Not removing key from BtreeMap so that failed withdrawals can still be
-						// tracked
-						Ok(())
-					} else {
-						// This allows us to ensure we do not have someone with an invalid account
-						Err(Error::<T>::InvalidWithdrawalIndex)
 					}
-				})
-				.is_ok(),
-				Error::<T>::InvalidWithdrawalIndex
```

Comment:
As mentioned [here](https://github.com/Polkadex-Substrate/Polkadex/pull/569#discussion_r992066811) I assume that's was the reason to wrap all into `ensure` and in that way catch possible error and basically `mapped` here to proper error. But ye, that solution is not intuitive...

Reply:
ensure can propagate only one error, included withing macros itself (usually used where only `bool` condition present). `?` propagates any given error - acting in same way but actually created for such scenarios :)